Preserve POSIX metadata

alokito opened the pull request:
Implements Issue awslabs/amazon-s3-tar-tool#23

When creating tar archives, copy POSIX metadata from S3 headers into TAR headers.
When extracting tar archives, set S3 headers using POSIX metadata from TAR headers.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

yanko7 commented:
Hello @alokito, thank you so much for your contribution! This looks great. Let me test it out and I'll merge it if all goes well.

alokito commented:
Thanks @yanko7 ! Two things

1. I'm a novice Go programmer, so I'd be grateful for any feedback no matter how nit-picky or small. Please do not hesitate to comment if there are any "code smells" or things that could be more idiomatically Go.
2. I have tested all three code paths (small files, large files, in memory) but only with a small test case. I am likely to test this against some multi-terabyte archives in the near future; we may want to hold off on merging until it passes the "real world" test.

In light of #2 I'll mark this as "draft" for now.

yanko7 commented on mem_concat.go:
```diff
@@ -25,7 +26,7 @@ func buildInMemoryConcat(ctx context.Context, client *s3.Client, objectList []*S
 
 	if estimatedSize < fileSizeMin {
-		data, err := tarGroup(ctx, client, objectList)
+		data, err := tarGroup(ctx, client, objectList, opts.PreservePOSIXMetadata)
```
I think it might be beneficial to pass the whole opts *S3TarS3Options object to tarGroup in case we need to pass more options in the future, that way the function type doesn't keep changing/growing.

alokito replied:
As you wish.

yanko7 commented on s3tar.go:
```diff
@@ -252,6 +288,18 @@ func concatObjAndHeader(ctx context.Context, svc *s3.Client, objectList []*S3Obj
 
+func fetchS3ObjectHead(ctx context.Context, svc *s3.Client, nextObject *S3Obj) *s3.HeadObjectOutput {
+	Debugf(ctx, "fetching head for %s/%s", *&nextObject.Bucket, *nextObject.Key)
+	head, err := svc.HeadObject(ctx, &s3.HeadObjectInput{
```
So one thing to consider as well is that every call to HeadObject counts as a GET request. This will essentially double the pricing to build the object. For example if you're trying to create an in-memory tarball with 1,000 Amazon S3 objects, you will need 1,000 GET requests to pull the data, then 1,000 requests to fetch the Head.

alokito replied:
Agreed. I don't think this is avoidable. This is why I decided to add the `--preserve-posix-metadata` command line argument, since there's no need to do these checks unless you are using FSX or Storage Gateway. I could move the check for `opts.PreservePOSIXMetadata` into `fetchS3ObjectHead` to make this more clear if you prefer.

yanko7 commented on mem_concat.go:
```diff
@@ -240,6 +241,10 @@ func tarGroup(ctx context.Context, client *s3.Client, objectList []*S3Obj) ([]by
 			Format:     tarFormat,
 		}
+		if preservePOSIXMetadata {
+			head := fetchS3ObjectHead(ctx, client, o)
```
The `GetObject` API call returns a Metadata dictionary as well. It might be worth modifying the `downloadS3Data` function so it also returns the header data that you need. Have you checked to see if it contains what is needed to presesrve the POSIXMetadata?

alokito replied:
Good catch! I have implemented this.

alokito commented:
@yanko7 I have run several validation tests in different scenarios, and I feel this is ready to be merged now. There were two additional changes since before:

- Although it is not documented, ctime is also populated by AWS FSX so I added code to preserve that
- When creating the TAR from parts, s3tar created a folder named output.tar/parts/ as well as a final file named output.tar. This confused FSX since the unix filesystem cannot have both a folder and file with the same name. I addressed this by changing the parts folder to output.tar.parts/, I hope this is acceptable.
